File: backend/src/crane/messages.py
from enum import Enum
from pydantic import BaseModel

class MessageType(str, Enum):
    CRANE_STATE = "crane_state"
    XYZ_POSITION = "xyz_position"

class BaseMessage(BaseModel):
    type: MessageType

class CraneStateTarget(BaseModel):
    swing: float
    lift: float
    elbow: float
    wrist: float
    gripper: float

class XYZPositionTarget(BaseModel):
    x: float
    y: float
    z: float

class CraneStateMessage(BaseMessage):
    type: MessageType = MessageType.CRANE_STATE
    target: CraneStateTarget

class XYZPositionMessage(BaseMessage):
    type: MessageType = MessageType.XYZ_POSITION
    target: XYZPositionTarget

File: frontend/src/types/messages.ts
import { CraneOrientation, XYZPosition, CraneState } from './crane';

export type MessageType = 'crane_state' | 'xyz_position';
export enum Status {
    MOVING = 'moving',
    STOPPED = 'stopped',
    ERROR = 'error'
}

export interface BaseMessage {
    type: MessageType;
    orientation: CraneOrientation;
}

export interface CraneStateMessage extends BaseMessage {
    type: 'crane_state';
    target: CraneState;
}

export interface XYZPositionMessage extends BaseMessage {
    type: 'xyz_position';
    target: XYZPosition;
}

export type WebSocketMessage = CraneStateMessage | XYZPositionMessage;

export interface Response {
    craneState?: CraneState;
    xyzPosition?: XYZPosition;
    targetState?: CraneState;
    targetXyzPostion?: XYZPosition;
    status: Status;
    errorMessage?: string;
} 